### ISSUE TITLE
Does your model generate acceptable audio at masked region?

### ISSUE THREAD

**Issue by bfs18** [lucidrains/e2-tts-pytorch#25]
I implemented an E2-TTS conditioning mechanism in my framework, which produces intelligible audio overall. However, the regions where ground truth Mel-spectrograms are provided result in unintelligible audio. I am curious whether this issue is due to a bug in my implementation or if your model encounters a similar problem. The problematic regions are specifically where the ground truth Mel-spectrograms are used in the following example.

gt
![media_images_valid_media_mel_in_870931_c84c6e16e846eeba17c4](https://github.com/user-attachments/assets/5d214082-3276-4cf0-a819-fcc370e88331)
pred
![media_images_valid_media_mel_hat_870931_9952248f6eba402d9536](https://github.com/user-attachments/assets/10963c7e-9075-43dd-80f6-663b0fec9035)

[gt wave](https://github.com/user-attachments/assets/bf581c4d-c4a6-4aca-8be8-5371c5b49402)

[pred wave](https://github.com/user-attachments/assets/96ab56d5-a671-4727-9543-79c7e9a811b0)

**Comment by lucidrains** [comment]
@bfs18 are you restoring the conditioning at each ode step [like this line of code?](https://github.com/lucidrains/e2-tts-pytorch/blob/main/e2_tts_pytorch/e2_tts.py#L673)

**Comment by lucidrains** [comment]
@bfs18 also, how is the text conditioning on your front? have you gotten over the hill and have it follow some text?

**Comment by bfs18** [comment]
Hi @lucidrains, I've solved the problem. I believe it was caused by the filled 0s introducing a harsh distribution shift to the Mel condition. So, I fill the [mean of the condition Mel instead](https://github.com/bfs18/rfwave/blob/f8a965700c8915e452e1c0dd009bafe036bda662/rfwave/dit.py#L424). My implementation differs from the paper in two ways: 1. I use phonemes rather than characters, and 2. I use a DIT backbone without the UNET skip connection. I obtained the results in just 3 days on the LJSpeech dataset. 


https://github.com/user-attachments/assets/33a77839-f53b-43a2-875d-d3ecde5b7a42

**Comment by lucidrains** [comment]
@bfs18 congrats on replicating the results! did you see accelerated convergence or did you pay a big price in those 3 days? regardless, all big data and compute will become small over time, and i guess architectural engineering for alignment is going the way of the dodo

**Comment by lucidrains** [comment]
@bfs18 somebody here has also tried phonemes and haven't seen anything remarkable, so does this mean you changed the text embed conditioning to be per layer using the DiT ada-ln and ada-zero?

**Comment by bfs18** [comment]
Hi @lucidrains  I used only an Nvidia 4090 for the training. There isn't a reliable metric to monitor the training progress; however, around 40% into the training, the audio starts to align with the text. The text and masked Mel spectrogram are fed exclusively into the first DiT layer in my code, as the DiT block is densely integrated with residual connections. 
Previously, I devoted my efforts to [RAD alignment](https://github.com/NVIDIA/radtts), but it did not yield better results than e2 conditioning."

**Comment by lucidrains** [comment]
@bfs18 thank you for sharing this! i'll try to strengthen the conditioning even more in this repo with some DiT specific circuits then.

**Comment by bfs18** [comment]
Hi @lucidrains I've separated the [e2_tts code from my project](https://github.com/bfs18/e2_tts), and it can generate intelligible results within one or two days of training on an Nvidia 4090 using the LJ Speech dataset. I didn't notice much difference between your implementation and mine. My codebase has been focused on TTS for a long time, so there might be some details I didn't notice but that work effectively. [This](https://drive.google.com/file/d/1plSXXMCOFlSBoMbKR7nAs0ramtxjTFKw/view?usp=sharing) is the test result after training for 500k steps.

**Comment by bfs18** [comment]
Hi @SWivid My mini-batch size is not fixed at 1. I utilize dynamic bucketing, where the batch is constructed from multiple audio samples to achieve a target total length.

**Comment by SWivid** [comment]
Thanks for reply @bfs18 ~
how is the time step, is it same within a mini-batch (with a target max frames) ?

**Comment by bfs18** [comment]
@SWivid No, each sample has its own time step.

**Comment by lucidrains** [comment]
@bfs18 that's awesome, thank you for sharing all this valuable information

i've added support for (english) phonemes as well as ada ln zero from the DiT paper, in the case those two details are important in any manner

**Comment by lucidrains** [comment]
@bfs18 one idea i've had, now that i'm more certain this paper is going to pan out, is to use the [mmdit](https://github.com/lucidrains/mmdit) block (with joint attention) from Esser and Rombach (proven out in both SD3 and Flux by now). this work would then be reduced down to a non-autoregressive VALL-E, but with separate parameters for the two modalities

**Comment by SWivid** [comment]
@lucidrains that is exactly what im trying to do
just one thing not sure, it sd3 rf predicting ε or v which has difference during inference

**Comment by lucidrains** [comment]
@SWivid i don't have enough experience with RF to really know which one is better yet

but i do have enough intuition with transformers to know that not overlapping the tokens is probably better. it still keeps with the "embarrassingly simple" theme

**Comment by lucidrains** [comment]
@SWivid do you think predicting noise is better? i should add that as an option [at this repo](https://github.com/lucidrains/rectified-flow-pytorch)

**Comment by lucidrains** [comment]
@SWivid actually, let me just try running a small experiment this morning

**Comment by SWivid** [comment]
@lucidrains I haven't reached a general conclusion yet

As described in the sd3 paper
> 3.1. Intuitively, however, the resulting velocity prediction target ε − x0 is more difficult for t in the middle of [0, 1]

I guess to predict noise could learn denoising over a wider range of time steps (in the middle)
while predicting v is to learn a tough work grasping the skeleton of voice, at time step very close to 0

everything is up to experiments

**Comment by lucidrains** [comment]
@SWivid it isn't wise these days to not trust Rombach and his team, as he has just an extraordinary track record. but yes, let me throw it into the RF repo for others to compare

**Comment by bfs18** [comment]
Hi @lucidrains Initially, I tried using mmdit, but the results were not satisfactory. Consequently, I switched to RAD alignment. In hindsight, perhaps I should have given mmdit more time. I believe the mmdit approach can be effective, as the e2 conditioning works well. In the mmdit block, it is intuitive for the left branch to receive the text and the right branch to receive the mel-spectrogram.

**Comment by lucidrains** [comment]
@bfs18 hmm yea, it would be strange for mmdit to underperform the current method

did you use separate parameters for text vs audio modalities?

**Comment by bfs18** [comment]
@lucidrains Positional embedding may play a role. I used RoPE on both text and mel without any elaboration. The RoPE of the text is close to the RoPE of the first few phonemes, which negatively influences alignment.

**Comment by lucidrains** [comment]
@bfs18 yes, that could be it

a better positional embedding would probably be dynamic positional bias, where the input also receives the type of modality as a small learned embedding

**Comment by lucidrains** [comment]
@bfs18 you did use separate parameters though or no? (feedforward and attention projections being different for text vs audio)

**Comment by bfs18** [comment]
@lucidrains I used separate parameters. It's like this.
![image](https://github.com/user-attachments/assets/a620de4e-71bf-433e-bc6f-47a0ecf68889)

**Comment by lucidrains** [comment]
@bfs18 nice! ok, will accept your anecdata :pray:

**Comment by bfs18** [comment]
@lucidrains @SWivid I didn't quite understand the sentence: `"Intuitively, however, the resulting velocity prediction target ε − x0 is more difficult for t in the middle of [0, 1], since for t = 0, the optimal prediction is the mean of p1, and for t = 1 the optimal prediction is the mean of p0."` when I was reading the paper. Could you explain it?  According to my understanding, the optimal velocity prediction should always be ε (Gaussian noise) − x0 (data). Why is it different at t=0 and t=1 in this sentence? Thank you in advance.

**Comment by lucidrains** [comment]
as you approach 0 or 1, you see either all noise, or all data

**Comment by lucidrains** [comment]
![results 7500](https://github.com/user-attachments/assets/c568f608-f518-415d-9799-a4031bf0635a)

@SWivid predict noise is looking good! took me a few tries; i figured out that the flow needs to be clipped in order for sampling to work well

**Comment by SWivid** [comment]
@lucidrains great!
have you take a term to ε-prediction loss ? as in paper:
![image](https://github.com/user-attachments/assets/8c8b110b-8a1b-4077-ac0c-c12e5171c5d9)

as i go through [rectified_flow](https://github.com/lucidrains/rectified-flow-pytorch/blob/main/rectified_flow_pytorch/rectified_flow.py) and diffusers repo of sd3,
the term is not considered in loss calcul

**Comment by lucidrains** [comment]
@SWivid no, i haven't done the proper loss weighting just yet

will get to it

**Comment by lucidrains** [comment]
going to close this issue, as it is technically resolved

**Comment by lucidrains** [comment]
@bfs18 i'm just curious, are you a researcher or a student? just so i get some context

**Comment by bfs18** [comment]
Hi @lucidrains I work at a company, and my job involves Text-to-Speech.

**Comment by lucidrains** [comment]
@bfs18 ah got it 👍

**Comment by cduk** [comment]
>I obtained the results in just 3 days on the LJSpeech dataset.

wait, what?! it only took 3 days to train on a single 4090?

**Comment by JohnHerry** [comment]
> Hi @SWivid My mini-batch size is not fixed at 1. I utilize dynamic bucketing, where the batch is constructed from multiple audio samples to achieve a target total length.

Is this dataset single speaker? what if alot of different speakers?
